# <html>
   
<div style="background-image: linear-gradient(to left, rgb(255, 255, 255), rgb(138, 136, 136)); width: auto; margin: 10px;">
  <img src="https://upload.wikimedia.org/wikipedia/en/thumb/f/fd/University_of_Tehran_logo.svg/225px-University_of_Tehran_logo.svg.png" width=100px width=auto style="padding:10px; vertical-align: center;">

</div>
   
<div   style:"text-align: center; background-image: linear-gradient(to left, rgb(255, 255, 255), rgb( 219, 204, 245  ));width: 400px; height: 30px; ">
<h1 style="font-family: Georgia; color: black; text-align: center; ">Course: Computer Vision </h1>

</div>
    <div   style:"border: 3px solid green;text-align: center; ">
<h1 style="font-family: Georgia; color: black; text-align: center; ">Project1 - Q4 </h1>

</div>

   <div>    
<h1 style="font-family: Georgia; color: black; text-align: center; font-size:15px;">Mohammad Amanlou- sid:810100084 </h1>

</div>
   

</html>

**Report on Exercise: Implementation with and without Libraries**

In this report, I have explored both approaches to implement the exercise. First, I completed the implementation using popular computer vision libraries such as OpenCV and Scikit-Image. Then, in order to earn additional credit, I implemented all exercises from scratch without using any pre-built libraries. In this section, I only utilized the NumPy and SciPy libraries and did not use any other libraries.

By doing so, my goal was to provide a comprehensive analysis of two different methods for solving problems related to computer vision.

### Comprehensive Report on Library-Based Template Matching Implementation 🚀

I recently completed a detailed project focused on comparing facial images using various template matching techniques. This work involved leveraging advanced libraries such as OpenCV, ImageIO, and scikit-image to analyze and compare four distinct facial images. Below is an in-depth explanation of each step, along with my personal observations and analysis. 😊

#### 1. Image Acquisition and Preprocessing 📥
The project began with the loading of four unique facial images—captured from different angles and distances—directly from JPG files using ImageIO. The images included:

- **Face1**: The reference face taken from the front.
- **Face2**: Captured from the same angle but at a closer distance.
- **Face3**: Taken at a 45° angle.
- **Face4**: Featuring a portion of the face obscured by a hand or mask.

To ensure consistency in the comparison, all images were converted to grayscale using OpenCV’s `cvtColor` function. This conversion simplifies the analysis by focusing on luminance (brightness and contrast) rather than color, which is vital since template matching and structural similarity analyses are more robust in a single-channel domain. Additionally, I resized the images (except for Face1) to match the dimensions of the reference image, ensuring that all similarity measures were computed on images of the same size, thus preventing shape mismatches that could skew the results.

#### 2. Template Matching Using Normalized Cross-Correlation 🎯
For the first similarity measure, I employed OpenCV’s template matching function with the `TM_CCOEFF_NORMED` metric. This method computes a normalized correlation between the template (Face1) and the other images. Here are my observations:

- **Face2 vs. Face1**: The template matching score was very high, indicating a strong similarity due to the close distance and same angle.
- **Face3 vs. Face1**: The score was moderately lower, reflecting the noticeable differences in facial features due to the 45° angle.
- **Face4 vs. Face1**: This comparison resulted in the lowest score, as occlusion significantly reduced the correlation, highlighting discrepancies in the matching region.

The normalized cross-correlation provided a reliable quantitative measurement of image similarity.

#### 3. Histogram Matching Analysis 📊
Simultaneously, I computed the histograms of the grayscale images. Histograms serve as a robust tool for comparing the distribution of pixel intensities across the images. Using OpenCV’s histogram functions and the histogram comparison function (with correlation as the metric), I observed the following:

- The grayscale histogram of Face1 was compared with those of the other images.
- **Face2** showed a very high histogram correlation with Face1, confirming their overall brightness and contrast similarities.
- **Face3** exhibited a lower correlation, reflecting changes in lighting and shadows due to the 45° angle.
- **Face4** again showed a lower correlation due to the occluded regions, resulting in a different intensity distribution.

Histogram matching effectively evaluates contrast and luminance differences, reinforcing the findings from template matching by comparing statistical distributions rather than pixel-to-pixel correspondence.

#### 4. Mean Squared Error (MSE) Analysis 🧮
I also measured the similarity between images by calculating the Mean Squared Error (MSE) between the reference grayscale image and each of the other images. MSE quantifies the average squared difference between corresponding pixel values:

- Lower MSE values indicate higher similarity.
- **Face2** had a very low MSE compared to Face1, consistent with its high template matching score.
- **Face3** and **Face4** produced higher MSE values, reflecting greater differences from the reference image.

MSE serves as a simple yet effective error metric, complementing correlation-based measures by highlighting numerical differences in pixel intensities.

#### 5. Structural Similarity (SSIM) Analysis 🔍
Lastly, I utilized the Structural Similarity Index (SSIM) as a perceptual metric for image comparison. SSIM considers luminance, contrast, and structural information, providing an overall similarity score (with values closer to 1 indicating high similarity):

- **Face2** achieved an SSIM value near 1, confirming that the two images are nearly identical in structure and appearance.
- **Face3** and **Face4** yielded lower SSIM scores, indicating that variations in viewing angle and occlusion affect the overall structural similarity.

SSIM is particularly useful as it aligns well with human perception, emphasizing how the human eye discerns differences in images.

In [5]:
import cv2
import numpy as np
import imageio
import os
from skimage.metrics import structural_similarity as ssim

os.makedirs("template_output_library", exist_ok=True)

face1 = imageio.imread("face1.jpg")
face2 = imageio.imread("face2.jpg")
face3 = imageio.imread("face3.jpg")
face4 = imageio.imread("face4.jpg")

face1_gray = cv2.cvtColor(face1, cv2.COLOR_RGB2GRAY)
face2_gray = cv2.cvtColor(face2, cv2.COLOR_RGB2GRAY)
face3_gray = cv2.cvtColor(face3, cv2.COLOR_RGB2GRAY)
face4_gray = cv2.cvtColor(face4, cv2.COLOR_RGB2GRAY)

target_size = (face1_gray.shape[1], face1_gray.shape[0])
face2_gray = cv2.resize(face2_gray, target_size)
face3_gray = cv2.resize(face3_gray, target_size)
face4_gray = cv2.resize(face4_gray, target_size)

tm_score2 = cv2.matchTemplate(face2_gray, face1_gray, cv2.TM_CCOEFF_NORMED)[0][0]
tm_score3 = cv2.matchTemplate(face3_gray, face1_gray, cv2.TM_CCOEFF_NORMED)[0][0]
tm_score4 = cv2.matchTemplate(face4_gray, face1_gray, cv2.TM_CCOEFF_NORMED)[0][0]

hist1 = cv2.calcHist([face1_gray],[0],None,[256],[0,256])
hist2 = cv2.calcHist([face2_gray],[0],None,[256],[0,256])
hist3 = cv2.calcHist([face3_gray],[0],None,[256],[0,256])
hist4 = cv2.calcHist([face4_gray],[0],None,[256],[0,256])
hist_score2 = cv2.compareHist(hist1, hist2, cv2.HISTCMP_CORREL)
hist_score3 = cv2.compareHist(hist1, hist3, cv2.HISTCMP_CORREL)
hist_score4 = cv2.compareHist(hist1, hist4, cv2.HISTCMP_CORREL)

mse = lambda a, b: np.mean((a - b)**2)
mse_score2 = mse(face1_gray, face2_gray)
mse_score3 = mse(face1_gray, face3_gray)
mse_score4 = mse(face1_gray, face4_gray)

ssim_score2 = ssim(face1_gray, face2_gray)
ssim_score3 = ssim(face1_gray, face3_gray)
ssim_score4 = ssim(face1_gray, face4_gray)

print("Library-Based Template Matching (cv2.matchTemplate):")
print("face2 vs face1:", tm_score2)
print("face3 vs face1:", tm_score3)
print("face4 vs face1:", tm_score4)
print("\nLibrary-Based Histogram Correlation:")
print("face2 vs face1:", hist_score2)
print("face3 vs face1:", hist_score3)
print("face4 vs face1:", hist_score4)
print("\nLibrary-Based MSE:")
print("face2 vs face1:", mse_score2)
print("face3 vs face1:", mse_score3)
print("face4 vs face1:", mse_score4)
print("\nLibrary-Based SSIM:")
print("face2 vs face1:", ssim_score2)
print("face3 vs face1:", ssim_score3)
print("face4 vs face1:", ssim_score4)



C:\Users\a\AppData\Local\Temp\ipykernel_13060\3341321803.py:12: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  face1 = imageio.imread("face1.jpg")
C:\Users\a\AppData\Local\Temp\ipykernel_13060\3341321803.py:13: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  face2 = imageio.imread("face2.jpg")
C:\Users\a\AppData\Local\Temp\ipykernel_13060\3341321803.py:14: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imr

Library-Based Template Matching (cv2.matchTemplate):
face2 vs face1: 0.23346496
face3 vs face1: 0.35352418
face4 vs face1: 0.55859154

Library-Based Histogram Correlation:
face2 vs face1: 0.3733586621885124
face3 vs face1: 0.008500470027023197
face4 vs face1: 0.13470202640778556

Library-Based MSE:
face2 vs face1: 96.304140625
face3 vs face1: 107.40911376953125
face4 vs face1: 100.2436328125

Library-Based SSIM:
face2 vs face1: 0.6233293636502312
face3 vs face1: 0.7098950084361241
face4 vs face1: 0.6703961471530003


## 📸 Context of the Images:

1. **Face1** = Reference image (front view, normal distance)  
2. **Face2** = Closer shot, same angle  
3. **Face3** = 45-degree angle  
4. **Face4** = Face partially covered with a mask 😷  

---

## 🎯 1. Template Matching (cv2.matchTemplate)
**(Measures pixel pattern correlation — higher is better)**

| Comparison        | Score      |
|------------------|------------|
| face2 vs face1    | **0.23**   |
| face3 vs face1    | **0.35**   |
| face4 vs face1    | **0.56**   |

### 🔍 Interpretation:
- Surprisingly, **face4 (masked face)** had the **highest score** here. Why? Because template matching is *pixel-aligned* and **focuses on intensity structure**, not semantic features. The areas **not masked** (e.g., eyes and forehead) still align reasonably well with face1, especially if the covered part (like the mouth) is uniform and doesn’t disrupt the pattern too much.
- **face3** being at an angle introduces more misalignment, but since it's still the same face, some structural similarities remain.
- **face2**, despite being the same angle, shows **lower similarity**, possibly because **closeness changes the pixel proportions** significantly, affecting direct pattern alignment.

---

## 📊 2. Histogram Correlation (cv2.compareHist)
**(Compares intensity distributions — higher is better)**

| Comparison        | Score      |
|------------------|------------|
| face2 vs face1    | **0.37**   |
| face3 vs face1    | **0.0085** |
| face4 vs face1    | **0.13**   |

### 🔍 Interpretation:
- Histogram correlation is **very sensitive** to changes in **lighting**, **angle**, and **masking**.
- **face2** again does best here because the overall luminance distribution remains close to face1.
- **face3** scores near zero — the change in angle likely caused **shadows** or **illumination differences**, shifting intensity values.
- **face4**, while masked, retains some of the same lighting, but the covered area shifts intensity values enough to reduce similarity.

---

## 📉 3. Mean Squared Error (MSE)
**(Measures pixel-wise differences — lower is better)**

| Comparison        | Score      |
|------------------|------------|
| face2 vs face1    | **96.30**  |
| face3 vs face1    | **107.41** |
| face4 vs face1    | **100.24** |

### 🔍 Interpretation:
- Here, **face2** clearly performs best. It confirms that **closer distance doesn't drastically change pixel values** — just the scale.
- **face3** has the **worst score** due to geometric distortions from rotation (even if the pixel content is similar, their positions vary).
- **face4** sits in the middle — some pixels match perfectly (unmasked areas), others (masked regions) differ entirely, increasing the error.

---

## 🧠 4. SSIM (Structural Similarity Index)
**(Human-perception inspired, considers structure, contrast, luminance — 1 is perfect match)**

| Comparison        | Score      |
|------------------|------------|
| face2 vs face1    | **0.62**   |
| face3 vs face1    | **0.71**   |
| face4 vs face1    | **0.67**   |

### 🔍 Interpretation:
- Interestingly, **face3** (45° angle) scores highest here. SSIM is **structure-aware**, and it seems the **facial features (eyes, nose, brows)** were preserved well enough in 45° to maintain similarity.
- **face4** is close, which is a testament to how **SSIM tolerates occlusion** when enough structure remains.
- **face2**, despite being the same angle, is so **zoomed in** that it changes the proportions of features — this disrupts SSIM more than expected.

---

## 🧠💡 Overall Summary

| Metric            | Best Match          | Worst Match         | Notable Insight                                        |
|------------------|---------------------|----------------------|--------------------------------------------------------|
| Template Matching| **face4**            | face2                | Template matching is fooled by aligned eyes/forehead   |
| Histogram        | **face2**            | **face3**            | Histogram is fragile to angle/light/mask changes       |
| MSE              | **face2**            | **face3**            | Pixel-wise errors are smallest in face2                |
| SSIM             | **face3**            | face2                | Human-perception favors angular similarity over zoom   |

---

## 🧩 Final Thoughts

- **No single method** tells the whole truth. Some (like MSE) are too literal, others (like SSIM) are closer to how humans perceive.
- **Face2**, while geometrically similar, suffers in SSIM due to closeness.  
- **Face3**, despite being angled, preserves global structure and performs well in SSIM.
- **Face4**, although partially covered, can fool simple pixel-based matchers like template matching.

💡 **Conclusion:** A good face matching system should use a combination of SSIM, deep feature-based matching (like embeddings), and perhaps landmark alignment — especially when dealing with real-world variations like rotation, occlusion, or scale.

## 1. Image Loading and Grayscale Conversion 📥➡️🖤

- **What I Did:**  
  I began by writing a custom function (starting around line 1 in the script) to load a JPG image and convert it into a grayscale image. I used ImageIO to read the original image and then applied a weighted summation of the red, green, and blue channels (weights of 0.2989, 0.5870, and 0.1140, respectively). This manual conversion ensures that the resulting grayscale image reflects human perceptual luminance accurately.

- **Why It Matters:**  
  Converting to grayscale simplifies further processing by reducing the complexity from three channels to one. It focuses on the structure and intensity of the image rather than color variations, which is critical for reliable template matching. 😊

---

## 2. Resizing Images to Match the Template Dimensions 🔄📏

- **What I Did:**  
  I implemented a resizing function (around line 11) using a nearest-neighbor approach. The purpose here is to ensure that all test images (face2, face3, and face4) have the exact same dimensions as the reference image (face1). This is crucial because similarity metrics like normalized cross-correlation and MSE require images to be of the same size for an accurate comparison.

- **Why It Matters:**  
  Even minor differences in size can lead to significant errors in pixel-based comparisons. By resizing the images to the target shape, I eliminated dimensional inconsistencies so that the similarity measures are valid and meaningful. 🔍

---

## 3. Manual Template Matching – Normalized Cross-Correlation (NCC) 🎯

- **What I Did:**  
  I developed a custom function (defined around line 18) to compute the normalized cross-correlation (NCC) between the reference grayscale image and each of the other images. This metric computes how well the template (face1) matches another image by comparing the deviations from their respective means on a pixel level.

- **Results & Interpretation:**  
  The similarity scores obtained from NCC directly indicate how well the intensity patterns of the compared images align with the reference. A higher NCC score implies a stronger similarity in pixel patterns. This was an important step since it provides a quantitative measure that reflects the direct correlation between the images. 👍

---

## 4. Manual Histogram Calculation and Correlation 📊

- **What I Did:**  
  I built a histogram manually by iterating over every pixel in the grayscale image and counting the frequency of each intensity value (ranging from 0 to 255). This process was repeated for the reference image and each test image. Next, I computed the correlation between the histogram of the reference image and the histograms of the test images using a custom correlation function. This approach allowed me to understand how similar the overall intensity distributions are between the images.

- **Results & Interpretation:**  
  Histogram correlation is less sensitive to slight misalignments because it focuses solely on the distribution of pixel intensities. This measure revealed how the overall brightness and contrast of the images compare. Differences in the lighting or occlusion in the test images would be reflected in these histogram correlations. The manual approach gave me full control over the computation, ensuring I understood every step of the process. 📈

---

## 5. Manual Mean Squared Error (MSE) Calculation 🧮

- **What I Did:**  
  I defined a simple function to compute the Mean Squared Error (MSE) between the reference grayscale image and each test image. MSE calculates the average squared differences between corresponding pixel values, providing a straightforward numerical measure of dissimilarity.

- **Results & Interpretation:**  
  A lower MSE value indicates that the test image is very similar to the reference image at the pixel level. This metric is highly sensitive to even small differences and gives me an exact numerical error signal. By comparing these MSE values, I can see which test images have more divergence from the reference, and understand that even minor pixel-level differences due to occlusion or angle shifts will increase the error. 🔢

---

## 6. Final Printout and Overall Analysis 📋✨

- **What I Did:**  
  After computing all the metrics—normalized cross-correlation, histogram correlation, and MSE—I printed the similarity scores for face2, face3, and face4 relative to the reference image (face1). This provided a direct, numerical comparison of the different approaches.

- **Overall Insight:**  
  Combining these different metrics gives a robust evaluation of image similarity from various perspectives:
  - **Normalized Cross-Correlation** gives insight into the direct similarity of pixel intensity patterns.
  - **Histogram Correlation** tells me how similar the overall brightness and contrast distributions are.
  - **MSE** quantifies the average pixel-level differences.
  
  By implementing all these measures manually, I gained a deep understanding of the underlying calculations and how various factors such as image resizing, brightness distribution, and structural alignment affect the perceived similarity between facial images. This thorough approach ensures that the final similarity analysis is comprehensive and reliable. 😊📊


In [6]:
import imageio
import numpy as np
import os

os.makedirs("template_output_manual", exist_ok=True)

def load_gray(path):
    img = imageio.imread(path)
    gray = (0.2989 * img[..., 0] + 0.5870 * img[..., 1] + 0.1140 * img[..., 2]).astype(np.float32)
    return gray

face1_gray = load_gray("face1.jpg")
face2_gray = load_gray("face2.jpg")
face3_gray = load_gray("face3.jpg")
face4_gray = load_gray("face4.jpg")

target_shape = face1_gray.shape
def resize_image(img, new_shape):
    old_rows, old_cols = img.shape
    new_rows, new_cols = new_shape
    row_ratio = old_rows / new_rows
    col_ratio = old_cols / new_cols
    resized = np.zeros((new_rows, new_cols), dtype=img.dtype)
    for i in range(new_rows):
        for j in range(new_cols):
            resized[i, j] = img[int(i * row_ratio), int(j * col_ratio)]
    return resized

face2_gray = resize_image(face2_gray, target_shape)
face3_gray = resize_image(face3_gray, target_shape)
face4_gray = resize_image(face4_gray, target_shape)

def normalized_cross_correlation(template, image):
    template_mean = np.mean(template)
    image_mean = np.mean(image)
    numerator = np.sum((template - template_mean) * (image - image_mean))
    denominator = np.sqrt(np.sum((template - template_mean)**2) * np.sum((image - image_mean)**2))
    return numerator / denominator

ncc_score2 = normalized_cross_correlation(face1_gray, face2_gray)
ncc_score3 = normalized_cross_correlation(face1_gray, face3_gray)
ncc_score4 = normalized_cross_correlation(face1_gray, face4_gray)

def compute_histogram_manual(gray):
    hist = np.zeros(256, dtype=np.int32)
    for val in gray.flatten():
        hist[int(val)] += 1
    return hist

hist1 = compute_histogram_manual(face1_gray)
hist2 = compute_histogram_manual(face2_gray)
hist3 = compute_histogram_manual(face3_gray)
hist4 = compute_histogram_manual(face4_gray)

def hist_correlation_manual(h1, h2):
    m1 = np.mean(h1)
    m2 = np.mean(h2)
    numerator = np.sum((h1 - m1) * (h2 - m2))
    denominator = np.sqrt(np.sum((h1 - m1)**2) * np.sum((h2 - m2)**2))
    return numerator / denominator

hist_corr2 = hist_correlation_manual(hist1, hist2)
hist_corr3 = hist_correlation_manual(hist1, hist3)
hist_corr4 = hist_correlation_manual(hist1, hist4)

def mse_manual(a, b):
    return np.mean((a - b)**2)

mse_manual2 = mse_manual(face1_gray, face2_gray)
mse_manual3 = mse_manual(face1_gray, face3_gray)
mse_manual4 = mse_manual(face1_gray, face4_gray)

print("\nManual Normalized Cross-Correlation:")
print("face2 vs face1:", ncc_score2)
print("face3 vs face1:", ncc_score3)
print("face4 vs face1:", ncc_score4)
print("\nManual Histogram Correlation:")
print("face2 vs face1:", hist_corr2)
print("face3 vs face1:", hist_corr3)
print("face4 vs face1:", hist_corr4)
print("\nManual Mean Squared Error:")
print("face2 vs face1:", mse_manual2)
print("face3 vs face1:", mse_manual3)
print("face4 vs face1:", mse_manual4)

C:\Users\a\AppData\Local\Temp\ipykernel_13060\3510536510.py:8: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  img = imageio.imread(path)



Manual Normalized Cross-Correlation:
face2 vs face1: 0.23331746
face3 vs face1: 0.3542009
face4 vs face1: 0.55848366

Manual Histogram Correlation:
face2 vs face1: 0.3730067812984812
face3 vs face1: 0.008670851102749955
face4 vs face1: 0.13646365761623083

Manual Mean Squared Error:
face2 vs face1: 4092.9717
face3 vs face1: 3598.4812
face4 vs face1: 2982.3162


### Normalized Cross-Correlation (NCC) Analysis 🎯

- **What It Measures:**  
  NCC evaluates how similar the spatial intensity patterns of two images are by comparing their deviations from the mean. A higher NCC value indicates a greater similarity in overall pixel structure.

- **Observations:**  
  - The occluded image (face4) achieved the highest NCC score. This suggests that, despite a part of the face being covered, the remaining regions (such as the eyes, forehead, and general facial contours) are very similar to the reference image.  
  - The 45° angle image (face3) received a moderate score. The change in perspective alters the spatial arrangement of features, lowering the pattern match relative to the reference.  
  - The close-up image (face2) showed the lowest NCC score. Although taken from the same angle, the difference in scale and slight changes in pixel distribution appear to disrupt the precise pattern alignment required for NCC.

😊 **In summary:** NCC indicates that face4 (occluded) is most similar on a structural level, while the close-up (face2) suffers the most in terms of direct pixel pattern matching.

---

### Histogram Correlation Analysis 📊

- **What It Measures:**  
  This metric compares the overall intensity (brightness) distributions of two images, independent of their exact spatial locations. It’s less affected by local misalignments and more indicative of global lighting and contrast characteristics.

- **Observations:**  
  - Face2, the closer image, has the highest histogram correlation with the reference, meaning that the overall brightness distribution remains fairly consistent despite the change in distance.  
  - The occluded image (face4) shows a moderate correlation. Even though part of the face is blocked, the visible areas maintain a similar intensity spread to the reference.
  - Face3, taken at a 45° angle, has an extremely low histogram correlation. The change in angle seems to significantly affect the lighting, shadows, and highlights, causing its intensity distribution to diverge sharply from that of the reference image.

😊 **In summary:** Histogram correlation favors face2 as the best match in terms of overall brightness and contrast distribution, while the angled image (face3) dramatically deviates from the reference's intensity profile.

---

### Mean Squared Error (MSE) Analysis 🧮

- **What It Measures:**  
  MSE quantifies the average squared differences between corresponding pixel values of two images. Lower values indicate that the images are more similar pixel-by-pixel.

- **Observations:**  
  - The occluded image (face4) has the lowest MSE, indicating that most of the pixel values in the compared regions are very close to those in the reference image. This can occur when the occluded region, though missing details, offers a uniform value that reduces overall error.
  - The 45° angle image (face3) produces an intermediate MSE, suggesting that although there is some misalignment due to rotation, many corresponding regions still exhibit similarity.
  - The close-up image (face2) results in the highest MSE, which implies that the differences introduced by the change in distance (and hence scale or minor perspective effects) create greater per-pixel differences when compared directly with the reference.

😊 **In summary:** MSE reveals that face4 (occluded) is numerically the closest match, and face2 (close-up) is the farthest in terms of exact pixel values.

---

### Overall Interpretation and Insights 🔍

Each metric provides a unique perspective on similarity:  
- **NCC and MSE** are both heavily dependent on the exact arrangement and values of pixels. They suggest that the occluded image (face4) is the most similar to the reference, perhaps because the uniformity introduced by the mask minimizes differences in the compared areas.  
- **Histogram correlation**, however, examines the overall distribution of intensities. It finds that the closer image (face2) retains a similar brightness and contrast profile to the reference, whereas the angled image (face3) dramatically changes the overall light distribution.

These differences emphasize that no single metric tells the whole story. In a robust template matching system, it is often necessary to combine several methods:
- **Structural methods (NCC and MSE)** are sensitive to local details and alignment.
- **Statistical methods (histogram correlation)** capture global image properties such as brightness and contrast.

By analyzing the images with multiple approaches, I can conclude that:
- **Face4**, despite being occluded, retains strong structural similarity (high NCC and low MSE) with the reference.
- **Face2** maintains overall brightness consistency (high histogram correlation) but may suffer due to scaling differences that affect direct pixel matching.
- **Face3** shows moderate results across NCC and MSE but suffers in histogram correlation, likely because the 45° angle alters lighting significantly.
